src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+			whereClauseList = list_make1(equalityExpr);
+
+			/* allocate column values and nulls arrays */
+			rel = heap_open(tableId, ExclusiveLock);
+			tupleDescriptor = RelationGetDescr(rel);
+			columnCount = tupleDescriptor->natts;
+			columnValues = palloc0(columnCount * sizeof(Datum));
+			columnNulls = palloc0(columnCount * sizeof(bool));
+
+			/*
+			 * We create a new memory context called tuple context, and read and write
+			 * each row's values within this memory context. After each read and write,
+			 * we reset the memory context. That way, we immediately release memory
+			 * allocated for each row, and don't bloat memory usage with large input
+			 * files.
+			 */
+			tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+												 "COPY Row Memory Context",
+												 ALLOCSET_DEFAULT_MINSIZE,
+												 ALLOCSET_DEFAULT_INITSIZE,
+												 ALLOCSET_DEFAULT_MAXSIZE);
+				
+			/* 
+			 * Construct hash table used for shardId->Connection mapping.
+			 * We can not use connection cache from connection.c used by GteConnection because
+			 * we need to establish multiple connections with each nodes: one connection per shard
+			 */
+			shard2conn = CreateShard2ConnectionHash();
+
+			/* init state to read from COPY data source */
+			copyState = BeginCopyFrom(rel, copyStatement->filename,
+									  copyStatement->is_program,
+									  copyStatement->attlist,
+									  copyStatement->options);
+				
+			/* TODO: handle binary mode */
+			if (copyState->binary) { 
+				elog(ERROR, "Copy in binary mode is not currently supported");
+			}
+
+			PG_TRY();
+			{
+				while (nextRowFound)
+				{
+					MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+					nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+					MemoryContextSwitchTo(oldContext);
+					
+					/* write the row to the shard */
+					if (nextRowFound)
+					{		
+						if (columnNulls[partitionColumn->varattno]) { 
+							ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+											errmsg("cannot copy row with NULL value "
+												   "in partition column")));
+						} else { 
+							Datum partitionValue = columnValues[partitionColumn->varattno];
+							rightConst->constvalue = partitionValue;
+							prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+							
+							foreach(shardIntervalCell, prunedList)
+							{
+								ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+								int64 shardId = shardInterval->id;
+								bool found;
+								
+								copyConn = (CopyConnection*)hash_search(shard2conn, &shardInterval->id, HASH_ENTER, &found);
+								if (!found) { 
+									List *finalizedPlacementList;
+									copyConn->nReplicas = 0;
+									copyConn->preparedMask = 0;
+									/* grab shared metadata lock to stop concurrent placement additions */
+									LockShardDistributionMetadata(shardId, ShareLock);
```

Comment:
Shard locks should be obtained in numerical order to avoid deadlock. You can maybe reuse AcquireExecutorShardLocks.

It would help the overall structure of the code to separate it into a planning and execution phase, where the planning phase generates tasks (maybe "streams") and the execution phase executes them. This would also make it more similar to existing pg_shard code.